### ISSUE TITLE
Fall 2018 Symposium Performance Talk Notes

### ISSUE THREAD

**Issue by tfboyd** [tensorflow/tensorflow#22201]
There were additional topics discussed that could not be turned into coherent notes after the fact.

## Action items
   * TensorFlow team to create a block diagram and then go deeper.  Goal is to help people who want to contribute plugins or extensions.
   * Make the NVIDIA Library to TF Version matrix more visible and consider better error messages.
   * Document tips and methods to use for large batch scaling with TensorFlow.  Possibly mention where this is proven and unproven to work.  This would include adding optimizers or wrapper to simplify usage.  Researchers are large labs want to scale but do not have the knowledge to do it quickly.

## General Questions and discussion results

   * Should TensorFlow default builds move forward more aggressively with newer versions of CUDA/cuDNN?
      * Internally Google moves slowly to new versions of CUDA as the verification process is long and often includes multiple patches with NVIDIA until all edge cases are covered.  The bar is extremely high for obvious reasons.
      * Group thought was it doesn’t matter. Situation would be improved with better error messages indicating what version is needed and making the matrix showing what NVIDIA libs are needed for each TensorFlow version.
   * Provide more models with good performance and accuracy in github.com/tensorflow/models.  Too many different versions and unsure which one to use.  Example:  There are 3+ ResNet models.
      * Contact tfboyd@ or comment in this thread if there is a specific model of interest and why.  MLPerf may improve this situation.
   * Auto regressive models with long dependency chains are hard to optimize by doing fusion by hand.
      * Consider trying XLA and providing feedback. An objective of XLA is to reduce the need for custom fusion.
   * [Ask] Distributing the input pipeline across servers.  Some work has been done and rumored to have occured in production but not trivial to setup.  Consider an RFP and/or reach out to the tf.data team.
   * TensorCores are only used if you are using tf.16.  You can do Pseudo FP16 with an envar that should be in TF 1.11 from NVIDIA as an experimental feature.  You still need to do loss scaling.

## Debugging and performance investigations
   * Hard to get timelines and profiler results.
      * New guide is on the way, not positive this will make it as easy as desired.
   * [Ask] Provide an interface to autotuning parameters such as intra and inter thread pools.
      * In hindsight, more information is needed to make this actionable.  Please add info to the comments.  

## Distributed compute (multi-gpu and multi-node)
   * All reduce API for multi-node and support MPI.
      * MPI would be useful for supercomputers where you need to use their MPI library to get good communication.
      * NCCL currently used by MirroredStrategy but only for multi-GPU.  TensorFlow’s own `ops` are used for multi-node all-reduce.

   * Document on how the distribution strategies is setup with the goal of showing how others can add their own solution and collective ops.  The API is the best place right now to figure this out.
   * [AI] TensorFlow team to create a block diagram and then go deeper.  Goal is to help people who want to contribute plugins or extensions.
   * Having a hard time with MPI.  People are using a newer versions of MPI
   * How do we plan to support fault tolerance? Will it allow dynamic addition and removal of machines to the collective. E.g. preemption but keep training with a smaller pool. Also increasing machines & batch size as training progresses. Need to adjust the learning rate, so maybe get a callback, especially when you are going to lose or have just lost a machine. Response to call back is starting with revised set of machines an new learning rate.
   *[AI] Document strategies for large batch training.  LARS looking to automatic learning rate, warmup rate and stuff.  

## Model parallelism and Data parallelism
   * Model parallelism is currently manual.  Work to automate it is on going and it can be phrased as an RL problem.  
   * Data parallelism supported via MirroredStrategy discussed today

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by tfboyd** [comment]
Updating soon.

**Comment by tfboyd** [comment]
There were additional topics discussed that could not be turned into coherent notes after the fact.

## Action items
   * TensorFlow team to create a block diagram and then go deeper.  Goal is to help people who want to contribute plugins or extensions.
   * Make the NVIDIA Library to TF Version matrix more visible and consider better error messages.
   * Document tips and methods to use for large batch scaling with TensorFlow.  Possibly mention where this is proven and unproven to work.  This would include adding optimizers or wrapper to simplify usage.  Researchers are large labs want to scale but do not have the knowledge to do it quickly.

## General Questions and discussion results

   * Should TensorFlow default builds move forward more aggressively with newer versions of CUDA/cuDNN?
      * Internally Google moves slowly to new versions of CUDA as the verification process is long and often includes multiple patches with NVIDIA until all edge cases are covered.  The bar is extremely high for obvious reasons.
      * Group thought was it doesn’t matter. Situation would be improved with better error messages indicating what version is needed and making the matrix showing what NVIDIA libs are needed for each TensorFlow version.
   * Provide more models with good performance and accuracy in github.com/tensorflow/models.  Too many different versions and unsure which one to use.  Example:  There are 3+ ResNet models.
      * Contact tfboyd@ or comment in this thread if there is a specific model of interest and why.  MLPerf may improve this situation.
   * Auto regressive models with long dependency chains are hard to optimize by doing fusion by hand.
      * Consider trying XLA and providing feedback. An objective of XLA is to reduce the need for custom fusion.
   * [Ask] Distributing the input pipeline across servers.  Some work has been done and rumored to have occured in production but not trivial to setup.  Consider an RFP and/or reach out to the tf.data team.
   * TensorCores are only used if you are using tf.16.  You can do Pseudo FP16 with an envar that should be in TF 1.11 from NVIDIA as an experimental feature.  You still need to do loss scaling.

## Debugging and performance investigations
   * Hard to get timelines and profiler results.
      * New guide is on the way, not positive this will make it as easy as desired.
   * [Ask] Provide an interface to autotuning parameters such as intra and inter thread pools.
      * In hindsight, more information is needed to make this actionable.  Please add info to the comments.  

## Distributed compute (multi-gpu and multi-node)
   * All reduce API for multi-node and support MPI.
      * MPI would be useful for supercomputers where you need to use their MPI library to get good communication.
      * NCCL currently used by MirroredStrategy but only for multi-GPU.  TensorFlow’s own `ops` are used for multi-node all-reduce.

   * Document on how the distribution strategies is setup with the goal of showing how others can add their own solution and collective ops.  The API is the best place right now to figure this out.
   * [AI] TensorFlow team to create a block diagram and then go deeper.  Goal is to help people who want to contribute plugins or extensions.
   * Having a hard time with MPI.  People are using a newer versions of MPI
   * How do we plan to support fault tolerance? Will it allow dynamic addition and removal of machines to the collective. E.g. preemption but keep training with a smaller pool. Also increasing machines & batch size as training progresses. Need to adjust the learning rate, so maybe get a callback, especially when you are going to lose or have just lost a machine. Response to call back is starting with revised set of machines an new learning rate.
   *[AI] Document strategies for large batch training.  LARS looking to automatic learning rate, warmup rate and stuff.  

## Model parallelism and Data parallelism
   * Model parallelism is currently manual.  Work to automate it is on going and it can be phrased as an RL problem.  
   * Data parallelism supported via MirroredStrategy discussed today

**Comment by byronyi** [comment]
I would be interested if @tfboyd could share more information on this particular event; I cannot find any information online apart from this issue.
